File: packages/skill-eval-python/src/skill_eval/runner/spinner.py
from __future__ import annotations

from rich.console import Console
from rich.progress import Progress, SpinnerColumn, TaskID, TextColumn
from rich.status import Status

from skill_eval.utils.string import truncate

_console = Console()


def _elapsed(ms: float) -> str:
    s = int(ms / 1000)
    if s < 60:
        return f"{s}s"
    m = s // 60
    return f"{m}m{s % 60}s"


class Spinner:
    def __init__(self) -> None:
        self._progress: Progress | None = None
        self._status: Status | None = None
        self._tasks: dict[str, TaskID] = {}
        self._total = 0
        self._completed = 0
        self._simple_message = ""
        self._global_start: float = 0
        self._task_labels: dict[str, str] = {}
        self._task_details: dict[str, str] = {}

    def start(self, total_or_message: int | str = 0) -> None:
        import time

        self._global_start = time.monotonic() * 1000
        if isinstance(total_or_message, str):
            self._total = 1
            self._simple_message = total_or_message
        else:
            self._total = total_or_message
            self._simple_message = ""
        self._completed = 0
        self._tasks.clear()
        self._task_labels.clear()
        self._task_details.clear()

        if self._simple_message:
            self._status = Status(self._simple_message, console=_console, spinner="dots")
            self._status.start()
        else:
            self._progress = Progress(
                SpinnerColumn(),
                TextColumn("[bold blue]{task.description}"),
                console=_console,
            )
            self._progress.start()

    def track(self, id: str, label: str) -> None:
        self._task_labels[id] = label
        self._task_details[id] = ""
        if self._progress is not None:
            task_id = self._progress.add_task(label, total=None)
            self._tasks[id] = task_id

    def untrack(self, id: str) -> None:
        self._task_labels.pop(id, None)
        self._task_details.pop(id, None)
        if id in self._tasks and self._progress is not None:
            self._progress.remove_task(self._tasks[id])
            del self._tasks[id]

    def detail(self, id: str, text: str) -> None:
        self._task_details[id] = text
        if id in self._tasks and self._progress is not None:
            label = self._task_labels.get(id, id)
            detail = truncate(text, 60)
            self._progress.update(
                self._tasks[id],
                description=f"{label} · {detail}" if detail else label,
            )

    def succeed(self, id: str, line: str) -> None:
        self._completed += 1
        self.untrack(id)
        _console.print(line)

    def stop(self) -> None:
        if self._status is not None:
            self._status.stop()
            self._status = None
        if self._progress is not None:
            self._progress.stop()
            self._progress = None

File: packages/skillet/src/eval/runner/spinner.ts
import pc from "picocolors";
import { truncate } from "../utils/string.js";

const FRAMES = ["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"];
const ANSI_PATTERN =
	// biome-ignore lint/suspicious/noControlCharactersInRegex: matches ANSI escape sequences
	/\u001B\[[0-9;?]*[ -/]*[@-~]/g;

function elapsed(ms: number): string {
	const s = Math.floor(ms / 1000);
	if (s < 60) return `${s}s`;
	const m = Math.floor(s / 60);
	return `${m}m${s % 60}s`;
}

function stripAnsi(text: string): string {
	return text.replace(ANSI_PATTERN, "");
}

function terminalColumns(): number {
	return Math.max(process.stdout.columns ?? 80, 20);
}

function wrappedRows(text: string): number {
	const columns = terminalColumns();
	return Math.max(
		1,
		text.split("\n").reduce((sum, line) => {
			const visibleLength = stripAnsi(line).length;
			return sum + Math.max(1, Math.ceil(Math.max(visibleLength, 1) / columns));
		}, 0),
	);
}

function fitLine(text: string): string {
	const columns = terminalColumns();
	const plain = stripAnsi(text);
	if (plain.length <= columns) return text;
	return `${plain.slice(0, Math.max(columns - 1, 1))}…`;
}

interface ActiveTask {
	label: string;
	detail: string;
	startTime: number;
}

type CompletionLine = string | ((completed: number, total: number) => string);

export class Spinner {
	private frameIdx = 0;
	private timer: ReturnType<typeof setInterval> | null = null;
	private isInteractive: boolean;
	private lines = 0;
	private globalStart = 0;
	private total = 0;
	private completed = 0;
	private active = new Map<string, ActiveTask>();
	private cycleIdx = 0;

	constructor() {
		this.isInteractive = process.stdout.isTTY ?? false;
	}

	private simpleMessage = "";

	start(totalOrMessage: number | string = 0) {
		if (typeof totalOrMessage === "string") {
			this.total = 1;
			this.simpleMessage = totalOrMessage;
		} else {
			this.total = totalOrMessage;
			this.simpleMessage = "";
		}
		this.completed = 0;
		this.globalStart = Date.now();

		if (!this.isInteractive) {
			if (this.simpleMessage) process.stdout.write(`  … ${this.simpleMessage}\n`);
			return;
		}

		this.render();
		this.timer = setInterval(() => {
			this.frameIdx = (this.frameIdx + 1) % FRAMES.length;
			if (this.frameIdx % 5 === 0) this.cycleIdx++;
			this.render();
		}, 80);
	}

	/** Register a task as actively running. */
	track(id: string, label: string) {
		this.active.set(id, { label, detail: "", startTime: Date.now() });
		if (!this.isInteractive) {
			process.stdout.write(`  … ${label}\n`);
		}
	}

	/** Remove a task without printing a result (e.g. before retry). */
	untrack(id: string) {
		this.active.delete(id);
	}

	/** Update the detail for a specific active task. */
	detail(id: string, text: string) {
		const task = this.active.get(id);
		if (task) task.detail = text;
	}

	/** Mark a task as completed and print its result line. */
	succeed(id: string, line: CompletionLine) {
		this.active.delete(id);
		this.completed++;
		if (this.isInteractive) this.clearLines();
		const renderedLine = typeof line === "function" ? line(this.completed, this.total) : line;
		process.stdout.write(`${renderedLine}\n`);
	}

	stop() {
		if (this.timer) clearInterval(this.timer);
		this.timer = null;
		if (this.isInteractive) this.clearLines();
	}

	private render() {
		this.clearLines();

		const frame = pc.cyan(FRAMES[this.frameIdx]);
		const time = pc.yellow(elapsed(Date.now() - this.globalStart));

		if (this.simpleMessage) {
			const spinnerLine = fitLine(
				`  ${frame} ${this.simpleMessage} ${pc.dim("(")}${time}${pc.dim(")")}`,
			);
			process.stdout.write(spinnerLine);
			this.lines = wrappedRows(spinnerLine);
			return;
		}

		const running = this.active.size;
		const status = `[${this.completed}/${this.total}] ${running} running`;
		const spinnerLine = fitLine(`  ${frame} ${status} ${pc.dim("(")}${time}${pc.dim(")")}`);

		const tasks = Array.from(this.active.values());
		if (tasks.length === 0) {
			process.stdout.write(spinnerLine);
			this.lines = wrappedRows(spinnerLine);
			return;
		}

		const shown = tasks.slice(0, 3);
		const detailLines = shown.map((t) => {
			const taskTime = pc.dim(elapsed(Date.now() - t.startTime));
			const detail = t.detail ? ` ${pc.dim("·")} ${pc.dim(truncate(t.detail, 60))}` : "";
			return fitLine(`    ${pc.dim("↳")} ${truncate(t.label, 40)} ${taskTime}${detail}`);
		});

		if (tasks.length > 3) {
			detailLines.push(fitLine(`    ${pc.dim(`  +${tasks.length - 3} more`)}`));
		}

		const block = `${spinnerLine}\n${detailLines.join("\n")}`;
		process.stdout.write(block);
		this.lines = wrappedRows(block);
	}

	private clearLines() {
		if (!this.lines) {
			process.stdout.write("\r");
			return;
		}
		const moveUp = this.lines > 1 ? `\x1b[${this.lines - 1}A` : "";
		process.stdout.write(`\r${moveUp}\x1b[J`);
		this.lines = 0;
	}
}
